BatchedTensor fallback: extended to support ops with multiple Tensor returns

zou3519 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #42700 Rename some BatchedTensorImpl APIs
* **#42628 BatchedTensor fallback: extended to support ops with multiple Tensor returns**

This PR extends the BatchedTensor fallback to support operators with
multiple Tensor returns. If an operator has multiple returns, we stack
shards of each return to create the full outputs.

For example, let's consider
```
tensor = torch.randn(B0, 10)
vmap(torch.var_mean)(torch.randn(B0, 10))
```
`torch.var_mean` returns two Tensors. The fallback path essentially does
the following:
- for each batch i, let ai, bi = torch.var_mean(tensor[i])
- return torch.stack([a0, a1, ..., a{B0}]), torch.stack([b0, b1, ..., b{B0}])

Test Plan:
- `pytest test/test_vmap.py -v`. Added a new test for an operator with
multiple returns (torch.var_mean).

Differential Revision: [D22957095](https://our.internmc.facebook.com/intern/diff/D22957095)

zou3519 committed:
BatchedTensor fallback: extended to support ops with multiple Tensor returns

This PR extends the BatchedTensor fallback to support operators with
multiple Tensor returns. If an operator has multiple returns, we stack
shards of each return to create the full outputs.

Test Plan:
- `pytest test/test_vmap.py -v`. Added a new test for an operator with
multiple returns (torch.var_mean).

[ghstack-poisoned]

zou3519 committed:
Update on "BatchedTensor fallback: extended to support ops with multiple Tensor returns"


This PR extends the BatchedTensor fallback to support operators with
multiple Tensor returns. If an operator has multiple returns, we stack
shards of each return to create the full outputs.

For example, let's consider
```
tensor = torch.randn(B0, 10)
vmap(torch.var_mean)(torch.randn(B0, 10))
```
`torch.var_mean` returns two Tensors. The fallback path essentially does
the following:
- for each batch i, let ai, bi = torch.var_mean(tensor[i])
- return torch.stack([a0, a1, ..., a{B0}]), torch.stack([b0, b1, ..., b{B0}])

Test Plan:
- `pytest test/test_vmap.py -v`. Added a new test for an operator with
multiple returns (torch.var_mean).

Differential Revision: [D22957095](https://our.internmc.facebook.com/intern/diff/D22957095)

[ghstack-poisoned]

ezyang commented on aten/src/ATen/BatchedFallback.cpp:
```diff
@@ -55,9 +55,9 @@ void batchedTensorForLoopFallback(const c10::OperatorHandle& op, torch::jit::Sta
+  TORCH_CHECK(num_returns >= 1,
               "Batching rule not implemented for ", schema, ". ",
-              "We do not yet support operations with multiple returns.");
+              "We do not support operations with no returns.");
```
shouldn't this be easy to support lol

zou3519 replied:
Yeah, this isn't hard to support. However it implies the operation does nothing (since it would have to have no returns and no alias annotations). I'll clarify that this means that the fallback doesn't support that

ezyang commented on aten/src/ATen/BatchedFallback.cpp:
```diff
@@ -105,8 +105,14 @@ void batchedTensorForLoopFallback(const c10::OperatorHandle& op, torch::jit::Sta
+  // [ a0, a1, a2, a3, b0, b1, b2, b3, c0, c1, c2, c3]
+  // This is so that we can call at::stack([a0...a3]), at::stack([b0...b3])
+  // more easily in the next step.
+  std::vector<Tensor> output_shards(num_batches * num_returns);
```
MatrixRef may be of interest here!

zou3519 replied:
I changed the indexing code to use MatrixRef! It's a nice way to abstract that behavior, thanks for pointing it out

zou3519 committed:
Update on "BatchedTensor fallback: extended to support ops with multiple Tensor returns"


This PR extends the BatchedTensor fallback to support operators with
multiple Tensor returns. If an operator has multiple returns, we stack
shards of each return to create the full outputs.

For example, let's consider
```
tensor = torch.randn(B0, 10)
vmap(torch.var_mean)(torch.randn(B0, 10))
```
`torch.var_mean` returns two Tensors. The fallback path essentially does
the following:
- for each batch i, let ai, bi = torch.var_mean(tensor[i])
- return torch.stack([a0, a1, ..., a{B0}]), torch.stack([b0, b1, ..., b{B0}])

Test Plan:
- `pytest test/test_vmap.py -v`. Added a new test for an operator with
multiple returns (torch.var_mean).

Differential Revision: [D22957095](https://our.internmc.facebook.com/intern/diff/D22957095)

[ghstack-poisoned]

zou3519 committed:
Update on "BatchedTensor fallback: extended to support ops with multiple Tensor returns"


This PR extends the BatchedTensor fallback to support operators with
multiple Tensor returns. If an operator has multiple returns, we stack
shards of each return to create the full outputs.

For example, let's consider
```
tensor = torch.randn(B0, 10)
vmap(torch.var_mean)(torch.randn(B0, 10))
```
`torch.var_mean` returns two Tensors. The fallback path essentially does
the following:
- for each batch i, let ai, bi = torch.var_mean(tensor[i])
- return torch.stack([a0, a1, ..., a{B0}]), torch.stack([b0, b1, ..., b{B0}])

Test Plan:
- `pytest test/test_vmap.py -v`. Added a new test for an operator with
multiple returns (torch.var_mean).

Differential Revision: [D22957095](https://our.internmc.facebook.com/intern/diff/D22957095)

[ghstack-poisoned]

zou3519 committed:
Update on "BatchedTensor fallback: extended to support ops with multiple Tensor returns"


This PR extends the BatchedTensor fallback to support operators with
multiple Tensor returns. If an operator has multiple returns, we stack
shards of each return to create the full outputs.

For example, let's consider
```
tensor = torch.randn(B0, 10)
vmap(torch.var_mean)(torch.randn(B0, 10))
```
`torch.var_mean` returns two Tensors. The fallback path essentially does
the following:
- for each batch i, let ai, bi = torch.var_mean(tensor[i])
- return torch.stack([a0, a1, ..., a{B0}]), torch.stack([b0, b1, ..., b{B0}])

Test Plan:
- `pytest test/test_vmap.py -v`. Added a new test for an operator with
multiple returns (torch.var_mean).

Differential Revision: [D22957095](https://our.internmc.facebook.com/intern/diff/D22957095)

[ghstack-poisoned]
